staticcheck/lint.go

```diff
@@ -3093,6 +3098,31 @@ func CheckRangeStringRunes(pass *analysis.Pass) (interface{}, error) {
 }
 
 func CheckSelfAssignment(pass *analysis.Pass) (interface{}, error) {
+	pure := pass.ResultOf[facts.Purity].(facts.PurityResult)
+
+	pureCall := func(e ast.Expr) bool {
+		call, ok := e.(*ast.CallExpr)
+		if !ok {
+			return false
+		}
+		if id, ok := call.Fun.(*ast.Ident); ok {
+			if fn, ok := pass.TypesInfo.ObjectOf(id).(*types.Func); ok {
```

Comment:
How about a pure function taking an argument that is impure or has side effects? Recursively? Also what about methods?

Reply:
Good catch regarding pure functions taking impure arguments. Addressed in 1a9134f.

Can you clarify what you mean about methods? Currently we just conservatively assume that all method calls may have side effects and the purity fact for methods is always false.